### PR TITLE
Bear.py: Include filename on which the bear failed

### PR DESCRIPTION
**Pull request by yash-nisar**
Mentions the absolute path of the file on which
 the local bear failed

Closes https://github.com/coala/coala/issues/2693

### Checklist

- [x] I have rebased properly. Please see [our tutorial on rebasing](http://coala.io/git#rebasing).
- [x] I have gone through the [commit guidelines](http://coala.io/commit) and I've followed them.
- [x] I have followed the [guidelines for the commit shortlog](http://coala.io/commit#shortlog).
- [x] I have followed the [guidelines for the commit body](http://coala.io/commit#commit-body).
- [x] I have included the issue URL with the 'Fixes' keyword if the commit fixes a *real bug* and the 'Closes' keyword if it adds a feature or enhancement. For details, check the [issue reference section in our commit guidelines](http://coala.io/commit#issue-reference).
- [x] I have [run all the tests](http://api.coala.io/en/latest/Developers/Executing_Tests.html) and they all pass. If any CI (below) is not green, please fix them. If GitMate issues appear you will have to amend your commits, pushing new commits on top is not sufficient!

**After you submit your pull request, DO NOT click the 'Update Branch' button**

### CODE REVIEW

File: coalib/bears/Bear.py [hunk 1]
```diff
@@ -249,8 +249,14 @@ def execute(self, *args, **kwargs):
             result = self.run_bear_from_section(args, kwargs)
             return [] if result is None else list(result)
         except (Exception, SystemExit):
-            self.warn('Bear {} failed to run. Take a look at debug messages'
-                      ' (`-V`) for further information.'.format(name))
+            if self.kind():
```
Comment (meetmangukiya) [inline coalib/bears/Bear.py]: This is incorrect, it'll be true for both local and global bears as well as linter bears making the else unreachable. Instead check it against the constants in the `coalib/bears/BEAR_KIND`.

Also add tests

Comment (yash-nisar) [inline coalib/bears/Bear.py]: Yeah, the tests are remaining.

Comment (yash-nisar) [inline coalib/bears/Bear.py]: However, I have replaced it with the condition `bear.kind() == BEAR_KIND.GLOBAL`. Is that what you implied to say ?

**Comment by yash-nisar** [comment]
Screwed up with the virualenv, idk how. :worried:

File: coalib/bears/Bear.py (hunk 1 above)
Comment (meetmangukiya) [inline coalib/bears/Bear.py]: Yeah, you are right.. I just assumed that BEAR_KIND.GLOBAL_BEAR would return something more informative, but it seems that we are using a it as a binary `0` is Local and `1` is global.

But I still suggest to use `self.kind() == BEAR_KIND.GLOBAL` as it is more readable and understandable then `if self.kind()`..

@Makman2 would you suggest the same ? ^^

Comment (Makman2) [inline coalib/bears/Bear.py]: yes please definitely do `if self.kind() == BEAR_KIND.XYZ`, relying on those values is bad as we use an enum which auto-generates those values^^

Comment (Makman2) [inline coalib/bears/Bear.py]: Btw `self.kind()` uses `NotImplemented` by default, maybe you should catch that^^

**Comment by yash-nisar** [comment]
Does this look good ? Please have a look at it when you are free :smile:  @Makman2

File: coalib/bears/Bear.py [hunk 2]
```diff
@@ -249,8 +250,19 @@ def execute(self, *args, **kwargs):
...
-                      ' (`-V`) for further information.'.format(name))
+            try:
+                if self.kind() == BEAR_KIND.GLOBAL:
+                    self.warn('Bear {} failed to run. Take a look '
+                              'at debug messages (`-V`) for further '
+                              'information.'.format(name))
+                else:
+                    self.warn('Bear {} failed to run on file {}. Take a look '
+                              'at debug messages (`-V`) for further '
+                              'information.'.format(name, args[0]))
```
Comment (Makman2) [inline coalib/bears/Bear.py]: actually I would rephrase the condition a bit:
```
if self.kind() == BEAR_KIND.LOCAL:
    # print file path
else:
    # normal message
```
conceptually you want to print filenames only on local bears, regardless what other types exist^^

File: coalib/bears/Bear.py [hunk 3]
```diff
@@ -249,8 +250,19 @@ def execute(self, *args, **kwargs):
...
+            except (Exception, SystemExit):
+                self.warn('Bear {} failed to run. Take a look '
+                          'at debug messages (`-V`) for further '
+                          'information.'.format(name))
```
Comment (Makman2) [inline coalib/bears/Bear.py]: huh why do you have the same try-catch block again?

File: tests/bears/BearTest.py [hunk 4]
```diff
@@ -129,6 +139,24 @@ def test_bad_bear(self):
         # debug message contains custom content, dont test this here
         self.queue.get()
 
+    def test_BadLocalBear(self):
+        self.uut = BadLocalBear(self.settings, self.queue)
```
Comment (Makman2) [inline tests/bears/BearTest.py]: why not giving directly `LocalBear` with a comment that because of no `run()` implementation it fails?

File: tests/bears/BearTest.py [hunk 5]
```diff
@@ -129,6 +139,24 @@ def test_bad_bear(self):
...
+        self.uut.execute('filename', 'file\n')
+        self.check_message(LOG_LEVEL.DEBUG)
+        self.check_message(LOG_LEVEL.WARNING,
+                           'Bear BadLocalBear failed to run on file filename. '
+                           'Take a look at debug messages (`-V`) for further '
+                           'information.')
+
+    def test_BadGlobalBear(self):
+        self.uut = BadGlobalBear(None, self.settings, self.queue)
```
Comment (Makman2) [inline tests/bears/BearTest.py]: same here, `GlobalBear` vs `BadGlobalBear`?

File: coalib/bears/Bear.py (hunk 3 above)
Comment (yash-nisar) [inline coalib/bears/Bear.py]: To catch the `NotImplementedError` of the kind() method by default.

Comment (Makman2) [inline coalib/bears/Bear.py]: but then catch for `NotImplementedError` + adapt the message :O

Comment (yash-nisar) [inline coalib/bears/Bear.py]: What should the adapted message be from your point of view ?

Comment (Makman2) [inline coalib/bears/Bear.py]: Actually I wouldn't catch it at all. If a bear without a kind makes it through this point, I don't know how to continue then. I mean without a kind coala doesn't know how to execute it (and me neither). A `NotImplementedError` with the stacktrace says everything relevant imo.

Comment (yash-nisar) [inline coalib/bears/Bear.py]: Yes, what you're saying is absolutely right. So what should the custom stacktrace message be like ? `The bear hasn't been implemented` or something similar ?

Comment (Makman2) [inline coalib/bears/Bear.py]: I would completely omit the try-catch^^

(Btw: Throwing an exception inside an except-block keeps the traceback alive, so don't `print`, but `raise` again :D)

Comment (yash-nisar) [inline coalib/bears/Bear.py]: Completely omitting the second `try catch` inside the `except` block will result in failure of some tests since the bear [here](https://github.com/coala/coala/blob/master/tests/bears/BearTest.py#L16) doesn't have a kind. So while executing [this](https://github.com/coala/coala/blob/master/tests/bears/BearTest.py#L121) test, when it enters the if condition ie. `self.kind() == BEAR_KIND.LOCAL`, it will throw an exception IIUC.

Comment (yash-nisar) [inline coalib/bears/Bear.py]: What do you suggest @Makman2 ?

File: tests/bears/BearTest.py [hunk 6]
```diff
@@ -129,6 +131,24 @@ def test_bad_bear(self):
         # debug message contains custom content, dont test this here
         self.queue.get()
 
+    def test_LocalBear(self):  # Fails because of no run() implementation
```
Comment (meetmangukiya) [inline tests/bears/BearTest.py]: Won't this failure be because of `NotImplementedError` ?

File: tests/bears/BearTest.py [hunk 7]
```diff
@@ -129,6 +131,24 @@ def test_bad_bear(self):
...
+        self.uut = LocalBear(self.settings, self.queue)
+        self.uut.execute('filename', 'file\n')
+        self.check_message(LOG_LEVEL.DEBUG)
+        self.check_message(LOG_LEVEL.WARNING,
+                           'Bear LocalBear failed to run on file filename. '
+                           'Take a look at debug messages (`-V`) for further '
+                           'information.')
+
+    def test_GlobalBear(self):  # Fails because of no run() implementation
```
Comment (meetmangukiya) [inline tests/bears/BearTest.py]: Even this one?

Comment (meetmangukiya) [inline tests/bears/BearTest.py]: Because in that case it gets into except block, then again the except `NotImplementedError` block, and the if else will never be executed ?

**Comment by meetmangukiya** [comment]
Rebase required

File: coalib/bears/Bear.py (hunk 3 above)
Comment (Makman2) [inline coalib/bears/Bear.py]: Try to give it a kind then :+1: :)

Comment (yash-nisar) [inline coalib/bears/Bear.py]: So, basically, the default should be replaced by a kind instead of a NotImplementedError ?

File: coalib/bears/Bear.py [hunk 8]
```diff
@@ -249,8 +250,19 @@ def execute(self, *args, **kwargs):
             result = self.run_bear_from_section(args, kwargs)
             return [] if result is None else list(result)
         except (Exception, SystemExit):
-            self.warn('Bear {} failed to run. Take a look at debug messages'
-                      ' (`-V`) for further information.'.format(name))
+            try:
+                if self.kind() == BEAR_KIND.LOCAL:
+                    self.warn('Bear {} failed to run on file {}. Take a look '
+                              'at debug messages (`-V`) for further '
+                              'information.'.format(name, args[0]))
```
Comment (Asnelchristian) [inline coalib/bears/Bear.py]: you could maybe, use a variable to store  this message, so that you don't repeat yourself

File: coalib/bears/Bear.py [hunk 9]
```diff
@@ -249,8 +250,19 @@ def execute(self, *args, **kwargs):
...
+                else:
+                    self.warn('Bear {} failed to run. Take a look '
+                              'at debug messages (`-V`) for further '
+                              'information.'.format(name))
+            except NotImplementedError:
+                self.warn('Bear {} failed to run. Take a look '
+                          'at debug messages (`-V`) for further '
+                          'information.'.format(name))
```
Comment (Asnelchristian) [inline coalib/bears/Bear.py]: Maybe you should clearly state that, the bear doesn't exist or that it hasn't been implemented

File: coalib/bears/Bear.py (hunk 8 above)
Comment (Asnelchristian) [inline coalib/bears/Bear.py]: and have something like ``.format(name, 'on file' + args[0])``

Comment (Makman2) [inline coalib/bears/Bear.py]: I would avoid assembling more "complex" grammatical structures using `format`. Though what makes sense is to append "Take a look at debug message ..." after the if I think^^

@Asnelchristian @yash-nisar opinions?

File: coalib/bears/Bear.py (hunk 9 above)
Comment (yash-nisar) [inline coalib/bears/Bear.py]: I removed the `try catch` since I gave a `DEFAULT` kind instead of `NotImplementedError`

File: coalib/bears/Bear.py (hunk 8 above)
Comment (yash-nisar) [inline coalib/bears/Bear.py]: I think we should keep 2 separate messages for better readability IMO @Makman2 @Asnelchristian

Comment (yash-nisar) [inline coalib/bears/Bear.py]: I think this pretty much satisfies everyone's opinion ? Please let me know if it doesn't. :smile:

File: coalib/bears/BEAR_KIND.py [hunk 10]
```diff
@@ -1,3 +1,3 @@
 from coalib.misc.Enum import enum
 
-BEAR_KIND = enum('LOCAL', 'GLOBAL')
+BEAR_KIND = enum('LOCAL', 'GLOBAL', 'DEFAULT')
```
Comment (Makman2) [inline coalib/bears/BEAR_KIND.py]: aaahh wait wait
what shall `DEFAULT`-kind mean?

I meant you shall add a kind to the bears defined in the tests :3

Comment (yash-nisar) [inline coalib/bears/BEAR_KIND.py]: I misinterpreted it. :(

Comment (yash-nisar) [inline coalib/bears/BEAR_KIND.py]: Is it okay now ? @Makman2

File: tests/bears/BearTest.py [hunk 11]
```diff
@@ -105,7 +116,7 @@ def tearDown(self):
     def test_simple_api(self):
         self.assertRaises(TypeError, TestBear, self.settings, 2)
         self.assertRaises(TypeError, TestBear, None, self.queue)
-        self.assertRaises(NotImplementedError, self.uut.kind)
+        self.assertEqual(BEAR_KIND.GLOBAL, self.uut.kind())
```
Comment (Makman2) [inline tests/bears/BearTest.py]: please create a new bear for it without an implementation for `kind`, this one explicitly tests the default behaviour^^

Comment (Makman2) [inline tests/bears/BearTest.py]: or just do `self.assertRaises(NotImplementedError, Bear.kind)` :+1: :)

**Comment by yash-nisar** [comment]
Looks good @Makman2 ?

File: tests/bears/BearTest.py [hunk 12]
```diff
@@ -129,6 +140,24 @@ def test_bad_bear(self):
         # debug message contains custom content, dont test this here
         self.queue.get()
 
+    def test_LocalBear(self):  # Fails because of no run() implementation
```
Comment (Makman2) [inline tests/bears/BearTest.py]: I would place the comment before `self.check_message(LOG_LEVEL.WARNING, ...)`, it somehow confuses me when I read it each time... sounds like the test case will always fail (which shouldn't happen, right :P)

Comment (Makman2) [inline tests/bears/BearTest.py]: same for the comment below

File: tests/bears/BearTest.py [hunk 13]
```diff
@@ -129,6 +140,24 @@ def test_bad_bear(self):
...
+        self.uut = LocalBear(self.settings, self.queue)
+        self.uut.execute('filename', 'file\n')
```
Comment (Makman2) [inline tests/bears/BearTest.py]: add an extension, makes the warning message below better to read ;)

File: tests/bears/BearTest.py (hunk 12 above)
Comment (Makman2) [inline tests/bears/BearTest.py]: Also the name of the testcase: You don't test the whole `LocalBear`, you test the aspect of printing the filename. Rename it to sth that fits better :+1:

Comment (Makman2) [inline tests/bears/BearTest.py]: same for testcase below

Comment (yash-nisar) [inline tests/bears/BearTest.py]: I renamed them as `test_print_filename_LocalBear` and `test_print_filename_GlobalBear` but we don't print filenames for the GlobalBear ? so should I rename it to something else or are the changes good ? @Makman2

Comment (Makman2) [inline tests/bears/BearTest.py]: then call it `test_print_no_filename_GlobalBear` :D

**Comment by Makman2** [comment]
unack b306990

see https://github.com/coala/coala/pull/3698#discussion_r100858385 :)

File: tests/bears/BearTest.py (hunk 12 above)
Comment (yash-nisar) [inline tests/bears/BearTest.py]: Done and rebased as well ! :D

**Comment by Makman2** [comment]
ack 1143282

**Comment by Makman2** [comment]
@rultor merge
